### ISSUE TITLE
Ambiguous parse for an integrity check

### ISSUE THREAD

**Issue by Gabriella439** [dhall-lang/dhall-lang#1025]
A user ran into an incorrect parse with the Haskell implementation, which I *think* is an issue with the standard, but I wanted to run this by others to make sure I'm not misinterpreting the standard before I submit a proposal to fix this.

The expression in question is essentially:

```dhall
https://example.com using ./headers.dhall sha256:3196e8dcb8ccc5f197223bedfda84d8ea66b37435e5d2735f4e6ae66bdf07dea
```

... from reading the standard grammar it appears that there are two valid parses for this expression:

* The first parse was what the user intended: to freeze the URL import:

  ```dhall
  (https://example.com using ./headers.dhall) sha256:3196e8dcb8ccc5f197223bedfda84d8ea66b37435e5d2735f4e6ae66bdf07dea
  ```

  I think we all agree that this interpretation is reasonable, otherwise there would be no way to freeze a URL import with custom headers

* The Haskell implementation selected the following alternative interpretation, which was to freeze the headers import:

  ```dhall
  https://example.com using (./headers.dhall sha256:3196e8dcb8ccc5f197223bedfda84d8ea66b37435e5d2735f4e6ae66bdf07dea)
  ```

  I think this is probably the unexpected interpretation.  If we really wanted to do this we could explicitly parenthesize

The issue with the grammar that permits the latter interpretation is that a `using` expression permits an `import-expression` as its right-hand side, which permits an integrity check.  I believe the fix we want is to create a new `import-type-expression` rule in the grammar which is the same as `import-expression` except not permitting an integrity check

**Comment by Gabriel439** [comment]
Also, they pointed out that this is the same issue as: https://github.com/dhall-lang/dhall-haskell/issues/1672

**Comment by Nadrieril** [comment]
I remember adding a unit test for this at some point. I'm not sure if either parse is more intuitive, I wouldn't mind rejecting both to force the user to choose

**Comment by ari-becker** [comment]
I'd like to expand by asking to consider the following questions:

```dhall
./local.dhall ? /usr/share/installed.dhall ? https://example.com/remote.dhall sha256:somehash
```

Is this the same as:

```dhall
./local.dhall sha256:somehash ? /usr/share/installed.dhall sha256:somehash ? https://example.com/remote.dhall sha256:somehash
```
(i.e., if the local import does not exist or does not meet the hash, then try installed, and if installed does not exist or does not meet the hash, then try the remote, and fail if remote does not meet the hash)

Or

```dhall
(./local.dhall ? /usr/share/installed.dhall ? https://example.com/remote.dhall) sha256:somehash
```
(i.e. if the local import exists, then use it, otherwise try the installed import, otherwise try the remote import, and only if one of them exist, test it against the hash. This is subtly different because if `local.dhall` exists, does not meet the hash, but the remote import does exist and meet the hash, then the import will fail, whereas in the first it will succeed)

Or

```dhall
./local.dhall ? /usr/share/installed.dhall ? (https://example.com/remote.dhall sha256:somehash)
```
in which case neither `local.dhall` nor `/usr/share/installed.dhall` are effectively annotated with a hash?

Personally I think that writing out the hash for each import is unergonomic and difficult to maintain - and that the intent of users in such cases is nearly always if not always to ensure that the final expression matches the hash.

**Comment by Gabriel439** [comment]
@ari-becker: The answer to your first question is that `a ? b ? c sha256:…` unambiguously parses as `a ? b ? (c sha256:…)`  This is because the left argument to an integrity check is necessarily an `import-type` which does not permit an arbitrary Dhall expression.

I think the latter question you raise probably should be covered in a separate issue as the focus of this issue is primarily on removing ambiguities from the grammar.

**Comment by ari-becker** [comment]
@Gabriel439 doesn't that mean that `https://example.com using (./headers.dhall sha256:3196e8dcb8ccc5f197223bedfda84d8ea66b37435e5d2735f4e6ae66bdf07dea)` is the unambiguously correct parse then? ;)

What I was trying to point out was that the parser is working as expected (i.e. unambiguously correct) and that the parser's behavior is counter-intuitive even when not considering the case of `using` and custom headers. It sounds to me like the problem is less one of ambiguity and more one of unintentional behavior. By fixing the order of operations by which a frozen hash is considered, it sounds to me like it fixes the user surprise, and in more cases, although maybe I'm not considering all of them.

**Comment by Gabriel439** [comment]
@ari-becker: `https://example.com using ./headers.dhall` is a valid `import-type` in the grammar, which is why the parse is ambiguous

**Comment by Nadrieril** [comment]
There was a proposal for allowing hashes anywhere. That would fix the issue with `?`. It doesn't say anything about `using` though

**Comment by Gabriel439** [comment]
Yeah, this is the relevant pull request for adding hashes anywhere: https://github.com/dhall-lang/dhall-lang/pull/668
